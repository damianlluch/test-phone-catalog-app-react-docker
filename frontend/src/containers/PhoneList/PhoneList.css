.PhoneList {
  display: flex;
  justify-content: center;
  align-items: center;
}

@media (max-width: 600px){
	.PhoneList{
		flex-direction: column;
	}
}
