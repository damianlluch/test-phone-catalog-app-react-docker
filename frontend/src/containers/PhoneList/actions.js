import axios from 'axios';
import C from '../../constants';

export const FETCH = '@Phones/FETCH';
export const fetch = () => ({
  type: FETCH,
});

export const FETCH_SUCCESS = '@Phones/FETCH_SUCCESS';
export const fetchSuccess = phones => ({
  type: FETCH_SUCCESS,
  payload: { phones },
});

export const FETCH_FAILURE = '@Phones/FETCH_FAILURE';
export const fetchFailure = error => ({
  type: FETCH_FAILURE,
  payload: { error },
});

export const RESET = '@Phones/RESET';
export const reset = () => ({
  type: RESET,
});

export const fetchPhones = () => dispatch => {
  dispatch(fetch());
  return axios
    .get(`${C.API_URL}/phones`)
    .then(res => dispatch(fetchSuccess(res.data)))
    .catch(error => dispatch(fetchFailure(error)));
};
