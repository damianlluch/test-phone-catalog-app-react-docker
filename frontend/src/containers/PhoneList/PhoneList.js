import React from 'react';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';
import PhoneListItem from './components/PhoneListItem';
import Spinner from '../../components/Spinner';
import Error from '../../components/Error';
import './PhoneList.css';

class PhoneList extends React.Component {
  componentDidMount() {
    this.props.fetchPhones();
  }

  componentWillUnmount() {
    this.props.reset();
  }

  renderList() {
    return this.props.items.map(phone => (
      <Link key={phone.id} to={`/phones/${phone.id}`}>
        <PhoneListItem
          key={phone.id}
          image={phone.details.image}
          title={phone.details.title}
          diagonal={phone.screen.diagonal}
          capacity={phone.hardware.capacity[0]}
        />
      </Link>
    ));
  }

  render() {
    const { items, loading, error } = this.props;

    if (loading) {
      return <Spinner />;
    }

    if (error) {
      return <Error error={error.message} />;
    }

    if (items) {
      return (
        <section className="PhoneList">
          {this.renderList()}
        </section>
      );
    }

    return null;
  }
}

PhoneList.propTypes = {
  items: PropTypes.array,
  loading: PropTypes.bool,
  error: PropTypes.object,
  fetchPhones: PropTypes.func,
  reset: PropTypes.func,
};

PhoneList.defaultProps = {
  items: [],
  loading: false,
  error: null,
  fetchPhones: null,
};

export default PhoneList;
