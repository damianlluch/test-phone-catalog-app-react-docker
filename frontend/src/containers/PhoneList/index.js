import { connect } from 'react-redux';
import PhoneList from './PhoneList';
import { fetchPhones, reset } from './actions';

const mapStateToProps = state => ({
  items: state.phones.items,
  loading: state.phones.loading,
  error: state.phones.error,
});

const mapDispatchToProps = dispatch => ({
  fetchPhones: () => dispatch(fetchPhones()),
  reset: () => reset(),
});

export default connect(mapStateToProps, mapDispatchToProps)(PhoneList);
