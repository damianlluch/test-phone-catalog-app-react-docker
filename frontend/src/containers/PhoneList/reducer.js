import { FETCH, FETCH_SUCCESS, FETCH_FAILURE, RESET } from './actions';

const INITIAL_STATE = {
  items: [],
  loading: false,
  error: null,
};

export default (state = INITIAL_STATE, action) => {
  switch (action.type) {
    case FETCH:
      return {
        ...state,
        loading: true,
        error: null,
      };

    case FETCH_SUCCESS:
      return {
        ...state,
        items: action.payload.phones,
        loading: false,
      };

    case FETCH_FAILURE:
      return {
        ...state,
        items: [],
        loading: false,
        error: action.payload.error,
      };

    case RESET:
      return {
        ...state,
        items: [],
        loading: false,
        error: null,
      };

    default:
      return state;
  }
};
