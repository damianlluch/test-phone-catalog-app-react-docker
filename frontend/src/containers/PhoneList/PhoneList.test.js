import React from 'react';
import renderer from 'react-test-renderer';
import { MemoryRouter } from 'react-router-dom';
import PhoneList from './PhoneList';

const phones = [
  {
    id: 'iphone',
    details: {
      title: 'Apple iPhone X',
      image:
        'https://d2giyh01gjb6fi.cloudfront.net/phone_front/0001/88/thumb_87957_phone_front_big.jpeg',
    },
    screen: {
      diagonal: '5.8',
    },
    hardware: {
      capacity: ['64 GB', '256 GB'],
    },
  },
  {
    id: 'galaxy',
    details: {
      title: 'Samsung Galaxy S9+',
      image:
        'https://d2giyh01gjb6fi.cloudfront.net/phone_front/0002/24/thumb_123965_phone_front_big.jpeg',
    },
    screen: {
      diagonal: '6.2',
    },
    hardware: {
      capacity: ['64 GB', '128 GB', '256 GB'],
    },
  },
];

describe('<PhoneList/>', () => {
  it('renders by default', () => {
    const tree = renderer.create(<PhoneList fetchPhones={() => {}} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders a list of phones', () => {
    const tree = renderer.create(<MemoryRouter><PhoneList fetchPhones={() => { }} items={phones} /></MemoryRouter>).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders an error', () => {
    const tree = renderer.create(<PhoneList fetchPhones={() => { }} error={{}} />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
