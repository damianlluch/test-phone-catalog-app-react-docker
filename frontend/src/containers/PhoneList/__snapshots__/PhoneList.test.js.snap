// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PhoneList/> renders a list of phones 1`] = `
<section
  className="PhoneList"
>
  <a
    href="/phones/iphone"
    onClick={[Function]}
  >
    <article
      className="PhoneListItem"
    >
      <img
        alt="Apple iPhone X"
        src="https://d2giyh01gjb6fi.cloudfront.net/phone_front/0001/88/thumb_87957_phone_front_big.jpeg"
      />
      <h2>
        Apple iPhone X
      </h2>
      <p>
        Diagonal: 5.8"
      </p>
      <p>
        Capacity: 64 GB
      </p>
      <p>
        <b>
          more specs...
        </b>
      </p>
    </article>
  </a>
  <a
    href="/phones/galaxy"
    onClick={[Function]}
  >
    <article
      className="PhoneListItem"
    >
      <img
        alt="Samsung Galaxy S9+"
        src="https://d2giyh01gjb6fi.cloudfront.net/phone_front/0002/24/thumb_123965_phone_front_big.jpeg"
      />
      <h2>
        Samsung Galaxy S9+
      </h2>
      <p>
        Diagonal: 6.2"
      </p>
      <p>
        Capacity: 64 GB
      </p>
      <p>
        <b>
          more specs...
        </b>
      </p>
    </article>
  </a>
</section>
`;

exports[`<PhoneList/> renders an error 1`] = `
<div
  className="Error"
>
  <h2 />
</div>
`;

exports[`<PhoneList/> renders by default 1`] = `
<section
  className="PhoneList"
/>
`;
